core/src/main/scala/org/apache/spark/sql/execution/ConnectionPool.scala

```diff
@@ -80,12 +81,13 @@ object ConnectionPool {
   def getPoolDataSource(id: String, props: Map[String, String],
       connectionProps: Properties, hikariCP: Boolean): DataSource = {
     // fast lock-free path first (the usual case)
-    val dsKey = idToPoolMap.get(id)
+    val username = props.getOrElse(Attribute.USERNAME_ALT_ATTR.toLowerCase, "")
+    val dsKey = idToPoolMap.get((id, username))
```

Comment:
Can create the lookup key tuple only once: `val lookupKey = id -> username` and use it in all the places.